dali/operators/math/expressions/broadcasting.cc

```diff
@@ -0,0 +1,242 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/math/expressions/broadcasting.h"
+#include "dali/operators/math/expressions/arithmetic_meta.h"
+
+namespace dali {
+
+void CheckNumSamples(span<const TensorListShape<>> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  int nsamples = shapes[0].num_samples();
+  for (int i = 1; i < shapes.size(); i++) {
+    DALI_ENFORCE(nsamples == shapes[i].num_samples());
+  }
+}
+
+template <typename Shape>
+int BroadcastNdimImpl(span<Shape> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  int ndim = shapes[0].size();
+  for (int i = 1; i < shapes.size(); i++) {
+    ndim = std::max(ndim, shapes[i].size());
+  }
+  return ndim;
+}
+
+int BroadcastNdim(span<const TensorShape<>> shapes) {
+  return BroadcastNdimImpl(shapes);
+}
+
+int BroadcastNdim(span<const TensorListShape<>> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  int ndim = shapes[0].sample_dim();
+  for (int i = 1; i < shapes.size(); i++) {
+    ndim = std::max(ndim, shapes[i].sample_dim());
+  }
+  return ndim;
+}
+
+template <typename Shape>
+void BroadcastShapeImpl(Shape& result, span<const Shape> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  if (shapes.size() == 1) {
+    result = shapes[0];
+    return;
+  }
+  int ndim = BroadcastNdim(shapes);
+  DALI_ENFORCE(result.size() == ndim);
+
+  for (int d = 0; d < ndim; d++) {
+    auto &res = result[result.size() - 1 - d];
+    res = 1;
+    for (int i = 0; i < shapes.size(); i++) {
+      auto &sh = shapes[i];
+      auto extent = d < sh.size() ? sh[sh.size() - 1 - d] : 1;
+      if (extent > 1 && extent != res) {
+        if (res > 1)
+          DALI_FAIL(make_string("Can't broadcast extents ", extent, " and ", res));
+        res = extent;
+      }
+    }
+  }
+}
+
+void BroadcastShape(TensorShape<>& result, span<const TensorShape<>> shapes) {
+  int ndim = BroadcastNdim(shapes);
+  result.resize(ndim);
+  BroadcastShapeImpl(result, shapes);
+}
+
+void BroadcastShape(TensorListShape<>& result, span<const TensorListShape<>> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  if (shapes.size() == 1) {
+    result = shapes[0];
+    return;
+  }
+  int ndim = BroadcastNdim(shapes);
+  CheckNumSamples(shapes);
+  int nsamples = shapes[0].num_samples();
+  result.resize(nsamples, ndim);
+
+  for (int sample_idx = 0; sample_idx < nsamples; sample_idx++) {
+    auto res_sample = result.tensor_shape_span(sample_idx);
+    for (int d = 0; d < ndim; d++) {
+      auto &res = res_sample[res_sample.size() - 1 - d];
+      res = 1;
+      for (int i = 0; i < shapes.size(); i++) {
+        auto sh = shapes[i].tensor_shape_span(sample_idx);
+        auto extent = d < sh.size() ? sh[sh.size() - 1 - d] : 1;
+        if (extent > 1 && extent != res) {
+          if (res > 1)
+            DALI_FAIL(make_string("Can't broadcast extents ", extent, " and ", res));
```

Comment:
We will probably need to be tad more verbose here, for the sake of the user. Exact coordinate, surrounding shape, etc.